go/store/prolly/tree/immutable_tree.go

```diff
@@ -29,137 +31,276 @@ import (
 
 const DefaultFixedChunkLength = 4000
 
-var ErrInvalidChunkSize = errors.New("invalid chunkSize; value must be > 1")
+var ErrInvalidChunkSize = errors.New("invalid chunkSize; value must be a multiple of 20")
 
-// buildImmutableTree writes the contents of |reader| as an append-only
+var chunkBufPool = sync.Pool{
+	New: func() any {
+		return new([][]byte)
+	},
+}
+
+func mustNewBlobBuilder(ns NodeStore, chunkSize int) *BlobBuilder {
+	b, _ := NewBlobBuilder(ns, chunkSize)
+	return b
+}
+
+// NewBlobBuilder writes the contents of |reader| as an append-only
 // tree, returning the root node or an error if applicable. |chunkSize|
 // fixes the split size of leaf and intermediate node chunks.
-func buildImmutableTree(ctx context.Context, r io.Reader, ns NodeStore, S message.Serializer, chunkSize int) (Node, error) {
-	if chunkSize < hash.ByteLen*2 || chunkSize > int(message.MaxVectorOffset)/2 {
-		// internal nodes must fit at least two 20-byte hashes
-		return Node{}, ErrInvalidChunkSize
+func NewBlobBuilder(ns NodeStore, chunkSize int) (*BlobBuilder, error) {
+	if chunkSize%hash.ByteLen != 0 {
+		return nil, ErrInvalidChunkSize
 	}
 
-	var levels [][]novelNode
-	var levelCnts []int
-	var finalize bool
-
-	// We use lookahead to check whether the reader has
-	// more bytes. The reader will only EOF when reading
-	// zero bytes into the lookahead buffer, but we want
-	// to know at the beginning of a loop whether we are
-	// finished.
-	lookahead := make([]byte, chunkSize)
-	lookaheadN, err := r.Read(lookahead)
-	if err != nil {
-		return Node{}, err
+	keys := make([][]byte, chunkSize/hash.ByteLen)
+	for i := range keys {
+		keys[i] = []byte{0}
+	}
+	return &BlobBuilder{
+		ns:         ns,
+		S:          message.NewBlobSerializer(ns.Pool()),
+		lChunkSize: chunkSize,
+		iChunkSize: chunkSize / hash.ByteLen,
+		buf:        make([]byte, chunkSize),
+		keys:       keys,
+		na:         &nodeArena{},
+		subtrees:   make([]uint64, chunkSize/hash.ByteLen),
+	}, nil
+}
+
+type BlobBuilder struct {
```

Comment:
Is the important thing here to write all the leaf chunks before we write any internal nodes and to write the internal nodes for each level contiguously? Or is the mostly allocation and duffcopy optimization?

This seems simpler to structure and maybe achieves less allocations if we build a chain of writers, one for each level and just call Write(reader) on it? Leaf writer just writes a leaf blob after reading into a buffer its constructed with. Internal writer just calls its child repeatedly, accumulating addrs into the buffer it's constructed with and flushing once it's full, returning to the addr to its caller.

A possible downside is this approach only writes runs of leaf nodes which fit in the first level internal node and doesn't write runs of internal nodes at all...

I have a small proof of concept which seems easy to follow and ~90 lines shorter but, not sure if it achieves the intent here...

Reply:
can you share your POC? it sounds like the regular chunker/old blob chunker algo

Reply:
Do you mean i should have kept the old DFS chunk building order, but have been smarter dividing a common backing slice for the level buffers? My main intent was to use the same two slices for all blobs, a byte buffer for leaf chunks, and an internal `[]hash.Hash` buffer for node addresses, though I wasn't even doing the address buffer properly. I need to write more go bench tests with increasing blob sizes, compare mallocs more closely.

Reply:
https://github.com/dolthub/dolt/pull/4830. Didn't optimize away the subtree counts here...

I just don't know which optimizations are the most important here...if we have a pool with the []byte, [][]byte, []uint64...I don't know if the DFS is important.

I guess another approach in #4830 would be to forgo the one-big-[]byte for the chain and just put every internal builder into the Pool and take out the number of you need for your height...

Reply:
thanks for highlighting the implementation forks, i also do not have a good sense of which are most important. i will benchmark the cpu/mem tradeoffs for the options on monday.